ARROW-17908: [Website] Arbitrarily Nested Data in Parquet and Arrow: Part 1: Introduction (#245)
apache/arrow-site: _posts/2022-10-01-arrow-parquet-encoding-part-1.md

As of version [20.0.0](https://crates.io/crates/arrow/20.0.0), released in August 2022, the Rust Arrow implementation for reading structured types is feature complete. Instructions for getting started can be found [here](https://docs.rs/parquet/latest/parquet/arrow/index.html) and feel free to raise any issues on our [bugtracker](https://github.com/apache/arrow-rs/issues).

In this series we will explain how Parquet and Arrow represent nested data, highlighting the similarities and differences between them, and giving a flavor of the practicalities of converting between the formats.

## Columnar vs Record-Oriented

First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as newline-delimited JSON (NDJSON), all the values for a given record are stored contiguously.

For example

```json
{"Column1": 1, "Column2": 2}
{"Column1": 3, "Column2": 4, "Column3": 5}
{"Column1": 5, "Column2": 4, "Column3": 5}
```

In a columnar representation, the data for a given column is instead stored contiguously

```text
Column1: [1, 3, 5]
Column2: [2, 4, 4]
Column3: [null, 5, 5]
```

> Aside from potentially yielding better data compression, a columnar layout can dramatically improve performance of certain queries. This is because laying data out contiguously in memory allows both the compiler and CPU to better exploit data parallelism. The specifics of [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) and [ILP](https://en.wikipedia.org/wiki/Instruction-level_parallelism) are well beyond the scope of this post, but the important takeaway is that processing large blocks of data without intervening conditional branches has substantial performance benefits.

Reviewer: ```suggestion
Aside from potentially yielding better data compression, a columnar layout can dramatically improve performance of certain queries. This is because laying data out contiguously in memory allows both the compiler and CPU to better exploit opportunities for parallelism. The specifics of [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) and [ILP](https://en.wikipedia.org/wiki/Instruction-level_parallelism) are well beyond the scope of this post, but the important takeaway is that processing large blocks of data without intervening conditional branches has substantial performance benefits.
```

I think this got lost in a rebase somewhere. ILP is not a form of data parallelism

---
layout: post
title: Arrow and Parquet Part 1: Primitive Types and Nullability

> date: "2022-10--01 00:00:00"

Author: ```suggestion
date: "2022-10-01 00:00:00"
```

author: tustvold, alamb
categories: [parquet, arrow]
---
<!--
{% comment %}
Licensed to the Apache Software Foundation (ASF) under one or more
contributor license agreements.  See the NOTICE file distributed with
this work for additional information regarding copyright ownership.
The ASF licenses this file to you under the Apache License, Version 2.0
(the "License"); you may not use this file except in compliance with
the License.  You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

## Introduction

> A long-running project within the [Rust Apache Arrow](https://github.com/apache/arrow-rs) implementation has been complete support for reading and writing arbitrarily nested parquet schema. This is a complex topic (as we will show), and we noticed a lack of available approachable technical information, and thus want to share our learnings with the community.

Reviewer: ```suggestion
A long-running project within the [Rust Apache Arrow](https://github.com/apache/arrow-rs) implementation has been complete support for reading and writing arbitrarily nested parquet schema. This is a complex topic, as we will show, and we noticed a lack of available approachable technical information, and thus wanted to share our learnings with the community.
```

> A long-running project within the [Rust Apache Arrow](https://github.com/apache/arrow-rs) implementation has been complete support for reading and writing arbitrarily nested parquet schema. This is a complex topic, as we will show, and we noticed a lack of available approachable technical information, and thus wanted to share our learnings with the community.

Reviewer 2: I am not native English speaker but `... has been complete support for ...` does not sound correct to me.

Reviewer: ```suggestion
A long-running project within the [Rust Apache Arrow](https://github.com/apache/arrow-rs) implementation has been adding complete support for reading and writing arbitrarily nested parquet schema. This is a complex topic, as we will show, and we noticed a lack of available approachable technical information, and thus wanted to share our learnings with the community.
```
Perhaps?

Reviewer 2: That's better!
Thanks, @tustvold !

Author: I tried to improve the wording in f210634631

[Apache Arrow](https://arrow.apache.org/) is an open, language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations. [Apache Parquet](https://parquet.apache.org/) is an open, column-oriented data file format designed for very efficient data encoding and retrieval.

> It is increasingly common for analytic systems to use Arrow to process data stored in Parquet files, so fast, efficient, and correct translation between them is a key building block.

Reviewer: ```suggestion
It is increasingly common for analytic systems to use Arrow to process data stored in Parquet files, and therefore fast, efficient, and correct translation between them is a key building block.
```

So fast made me think of "it's so fast", as opposed to introducing a subordinate clause

Historically analytic processing primarily focused on querying data with a tabular schema, that is one where there are a fixed number of columns, and each row contains a single value for a given column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.

As of version [20.0.0](https://crates.io/crates/arrow/20.0.0), released in August 2022, the Rust implementation is feature complete. Instructions for getting started can be found [here](https://docs.rs/parquet/latest/parquet/arrow/index.html) and feel free to raise any issues on our [bugtracker](https://github.com/apache/arrow-rs/issues).

> In this series article we will explain how Parquet and Arrow represent nested data, highlighting the similarities and differences between them, and giving a flavor of the practicalities of supporting reading and writing between them.

Reviewer: ```suggestion
In this series we will explain how Parquet and Arrow represent nested data, highlighting the similarities and differences between them, and giving a flavor of the practicalities of supporting reading and writing between them.
```

First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as JSON, all the values for a given record are stored contiguously.

For example

```json
{“Column1”: 1, “Column2”: 2}
{“Column1”: 3, “Column2”: 4, “Column3”: 5}
{“Column1”: 5, “Column2”: 4, “Column3”: 5}
```

In a columnar representation, the data for a given column is instead stored contiguously

```text
Column1: [1, 3, 5]
Column2: [2, 4, 4]
Column3: [null, 5, 5]
```

Aside from potentially yielding better data compression, a columnar layout can dramatically improve performance of certain queries. This is because laying data out contiguously in memory allows both the compiler and CPU to better exploit opportunities to process the data in parallel. The specifics of [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) and [ILP](https://en.wikipedia.org/wiki/Instruction-level_parallelism) are well beyond the scope of this post, but the important takeaway is that processing large blocks of data without intervening conditional branches has substantial performance benefits.

## Parquet vs Arrow

> Parquet and Arrow are complementary technologies, and they make some different design tradeoffs. In particular, Parquet is a storage format designed for maximum space efficiency, whereas Arrow is an in-memory format intended to be operated on by vectorized computational kernels.

Reviewer: ```suggestion
Parquet and Arrow are complementary technologies, but they make some different design tradeoffs. In particular, Parquet is a storage format designed for maximum space efficiency, whereas Arrow is an in-memory format intended to be operated on by vectorized computational kernels.
```

The major distinction is that arrow provides O(1) random access lookups to any array index, whilst parquet does not. In particular, Parquet uses [dremel record shredding](https://akshays-blog.medium.com/wrapping-head-around-repetition-and-definition-levels-in-dremel-powering-bigquery-c1a33c9695da), [variable length encoding schemes](https://github.com/apache/parquet-format/blob/master/Encodings.md), and [block compression](https://github.com/apache/parquet-format/blob/master/Compression.md) to drastically reduce the data size, but these techniques come at the loss of performant random access lookups.

A common pattern that plays to each technologies strengths, is to stream data from a compressed representation, such as parquet, in thousand row batches in the arrow format, process these batches individually, and accumulate the results in a more compressed representation. This benefits from the ability to efficiently perform computations on arrow data, whilst keeping memory requirements in check, and allowing the computation kernels to be agnostic to the encodings of the source and destination.

> **Arrow is primarily an interchange format, whereas Parquet is a storage format.**

Reviewer 3: can we enhance it as "arrow is in-memory/interchange format"

Author: Fixed in 6b14b13d81

## Non-Nullable Primitive Column

Let us start with the simplest case of a non-nullable list of 32-bit signed integers.

In arrow this would be represented as a `PrimitiveArray`, which would store them contiguously in memory

```text
┌─────┐
│  1  │
├─────┤
│  2  │
├─────┤
│  3  │
├─────┤
│  4  │
└─────┘
Values
```

Parquet has multiple different encodings that it can use for integers types, the exact details of which are beyond the scope of this post, but broadly speaking it will encode the data as one or more pages containing the integers

```text
┌─────┐
│  1  │
├─────┤
|  2  │
├─────┤
│  3  │
├─────┤
│  4  │
└─────┘
 Data

 DataPage
```

# Nullable Primitive Column

> Now let us consider the case of a nullable column, where some of the values might have the special sentinel value `NULL` that designates “this value is unknown.”

Reviewer: ```suggestion
Now let us consider the case of a nullable column, where some of the values might have the special sentinel value `NULL` that designates “this value is unknown".
```

> In Arrow nulls are stored separately from the values in the form of a validity bitmask, with arbitrary data in the corresponding positions in the values buffer.

Reviewer 3: can add link to https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps

Author: in 43860e33b9

```text
┌─────┐   ┌─────┐
│  1  │   │  1  │
├─────┤   ├─────┤
│  0  │   │ ??  │
├─────┤   ├─────┤
│  1  │   │  3  │
├─────┤   ├─────┤
│  1  │   │  4  │
├─────┤   ├─────┤
│  0  │   │ ??  │

> └─────┘   └─────┘

Reviewer 3: This diagram is giving a perception that bitmap is also same size (like an integer), which I assume is not the case. It's a bitmap with corresponding bit set. I do not have any bright ideas to enhance this, perhaps a note.

Author: In 43860e33b9

Validity   Values
```

> In Parquet the validity information is also stored separately from the values, however, instead of being encoded as a validity bitmask it is encoded as a list of 16-bit integers called definition levels. These will be expanded upon later, but for now a definition level of 1 indicates a valid value, and 0 a null value. Unlike arrow, nulls are not encoded in the list of values

Reviewer: ```suggestion
In Parquet the validity information is also stored separately from the values, however, instead of being encoded as a validity bitmask it is encoded as a list of 16-bit integers called definition levels. These will be expanded upon in the next post, but for now a definition level of 1 indicates a valid value, and 0 a null value. Unlike arrow, nulls are not encoded in the list of values
```

distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
{% endcomment %}
-->

## Introduction

We recently completed a long-running project within [Rust Apache Arrow](https://github.com/apache/arrow-rs) to complete support for reading and writing arbitrarily nested Parquet and Arrow schemas. This is a complex topic, and we encountered a lack of approachable technical information, and thus wrote this blog to share our learnings with the community.

[Apache Arrow](https://arrow.apache.org/) is an open, language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations. [Apache Parquet](https://parquet.apache.org/) is an open, column-oriented data file format designed for very efficient data encoding and retrieval.

It is increasingly common for analytic systems to use Arrow to process data stored in Parquet files, and therefore fast, efficient, and correct translation between them is a key building block.

> Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for a each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.

Reviewer 4: ```suggestion
Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can seem antiquated and unnecessarily limiting.
```

Reviewer 4: (just removing the extra `a`)

Reviewer 4: but, now that I read the next paragraph, it feels like a cliffhanger to say that the schema limitation "can seem antiquated and [limiting]." and not address whether it is/isn't antiquated/limiting or where you might discuss such a topic.

If that's not covered in the blog post, it seems better to just say that structured document formats are an alternative that do not expect a schema describing the structure.

Reviewer 4: but I don't think using XML/JSON requires flattening of a source schema. Or, maybe you're thinking in the direction of going from JSON/XML -> tabular?

I think I would prefer them to both be introduced neutrally. Historically tabular schemas have been used, but JSON and XML are 2 popular alternate formats that do not need a schema.

Reviewer: > Or, maybe you're thinking in the direction of going from JSON/XML -> tabular?

Yes that is precisely the transformation that has been a repeated pain point in my experience of data pipelines

Reviewer: What do you think of

```suggestion
Historically analytic processing primarily focused on querying data with a tabular schema, where there are a fixed number of columns, and each row contains a single value for each column. However, with the increasing adoption of structured document formats such as XML, JSON, etc…, this schema limitation can be frustrating for users, as it necessitates often non-trivial data transformation to first flatten the document data into a tabular format.
```

Reviewer 4: That seems reasonable.

IMO, the use of "Historically" and "schema limitation" makes it seem like you're trying to suggest that schemas can be frustrating. Maybe this, or some things from it, could be an improvement:

```suggestion
Historically, analytic processing primarily focused on querying data with a tabular schema. Recently, there has been an increasing adoption of structured document formats such as XML or JSON. For structured document formats, specifying a schema can seem limiting. For tabular data, non-trivial data transformations to flatten document data is often required. For these reasons, translating between the two can be frustrating for users.
```

Reviewer: I think that is more confusing as it seems to suggest that document formats don't have schema??

Author: Thanks for sticking with this -- nice improvement

As of version [20.0.0](https://crates.io/crates/arrow/20.0.0), released in August 2022, the Rust Arrow implementation for reading structured types is feature complete. Instructions for getting started can be found [here](https://docs.rs/parquet/latest/parquet/arrow/index.html) and feel free to raise any issues on our [bugtracker](https://github.com/apache/arrow-rs/issues).

In this series we will explain how Parquet and Arrow represent nested data, highlighting the similarities and differences between them, and giving a flavor of the practicalities of converting between the formats.

## Columnar vs Record-Oriented

> First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as JSON, all the values for a given record are stored contiguously.

Reviewer 4: I don't think JSON or other formats are inherently record-oriented or column-oriented. Also, depending on how pedantic you want to be, "column-oriented" or "record-oriented" are probably better described as data layouts; not "formats". I believe other sources describe them similarly (I just checked [this wiki page](https://en.wikipedia.org/wiki/Column-oriented_DBMS)).

I think it's better to change the description to say "layout," which may be defined along the lines of "how the physical bytes are stored in memory or on storage." You can keep these examples as is, but maybe clarify in the direction of, "When using a record-oriented layout, values for a given record are stored contiguously. Here's how we might represent this data using the JSON format."

Ultimately, I understand being terse, but it should be clear to the reader that a column-oriented layout can be represented in JSON because it is a format with no constraints on columns or records:
```json
{
    "Column1": [1, 3, 5],
    "Column2": [2, 4, 4],
    "Column3": [null, 5, 5]
}
```

Reviewer 4: (also, a few paragraphs down "layout" is used, so yeah)

Reviewer: ```suggestion
First, it is necessary to take a step back and discuss the difference between columnar and record-oriented data formats. In a record oriented data format, such as newline-delimited JSON (NDJSON), all the values for a given record are stored contiguously.
```

I was specifically referring to ndjson here. Does that allay your concerns?

Reviewer: njson is a record-oriented format, parquet is a column-oriented format, is this incorrect?

Reviewer 4: I see. I didn't know about NDJSON specifically assuming records. Its format still doesn't preclude a column-oriented representation, but its assumption does.

then, you're right and this can be ignored.

For example

```json

> {"Column1": 1, "Column2": 2}
> {"Column1": 3, "Column2": 4, "Column3": 5}
> {"Column1": 5, "Column2": 4, "Column3": 5}

Reviewer 4: ```suggestion
[
    {"Column1": 1, "Column2": 2},
    {"Column1": 3, "Column2": 4, "Column3": 5},
    {"Column1": 5, "Column2": 4, "Column3": 5}
]
```

Isn't it better to have valid JSON? This is not important, though, so feel free to ignore

Reviewer: This is meant to be ndjson, not JSON in the broad

Reviewer 4: I see, then this is fine.

The major distinction is that Arrow provides `O(1)` random access lookups to any array index, whilst Parquet does not. In particular, Parquet uses [dremel record shredding](https://akshays-blog.medium.com/wrapping-head-around-repetition-and-definition-levels-in-dremel-powering-bigquery-c1a33c9695da), [variable length encoding schemes](https://github.com/apache/parquet-format/blob/master/Encodings.md), and [block compression](https://github.com/apache/parquet-format/blob/master/Compression.md) to drastically reduce the data size, but these techniques come at the loss of performant random access lookups.

A common pattern that plays to each technologies strengths, is to stream data from a compressed representation, such as Parquet, in thousand row batches in the Arrow format, process these batches individually, and accumulate the results in a more compressed representation. This benefits from the ability to efficiently perform computations on Arrow data, whilst keeping memory requirements in check, and allowing the computation kernels to be agnostic to the encodings of the source and destination.

> **Arrow is primarily an in-memory format, whereas Parquet is a storage format.**

Reviewer 4: Could be overly compicated, but perhaps change to:
"Arrow is designed for efficient in-memory operations, whereas Parquet is designed for minimal footprint on storage devices."

Otherwise, it makes arrow file format potentially confusing?

Reviewer: I think I would rather emphasise that arrow is not designed as a storage format, it does exist, as alluded to by "primarily", but the major motivation for this appears to be spilling or disk-based IPC.

Reviewer 4: okay, sounds good.

(I do think *most* storage is "disk-based IPC", and I suspect with NVRAM this blurs even further)

## Non-Nullable Primitive Column

Let us start with the simplest case of a non-nullable list of 32-bit signed integers.

In Arrow this would be represented as a `PrimitiveArray`, which would store them contiguously in memory

```text
┌─────┐
│  1  │
├─────┤
│  2  │
├─────┤
│  3  │
├─────┤
│  4  │
└─────┘
Values
```

Parquet has multiple [different encodings](https://parquet.apache.org/docs/file-format/data-pages/encodings/) that may be used for integer types, the exact details of which are beyond the scope of this post. Broadly speaking the data will be stored in one or more [*DataPage*](https://parquet.apache.org/docs/file-format/data-pages/)s containing the integers in an encoded form

```text
┌─────┐
│  1  │
├─────┤
|  2  │
├─────┤
│  3  │
├─────┤
│  4  │
└─────┘
 Data

>  DataPage

Reviewer 4: also, you can probably have the examples in 1 excerpt instead of 2 since they are narrow.

>  DataPage

Reviewer: I would also vote for removing this detail

>  DataPage

Author: removed in b317b4328c
